apollo-runtime/src/main/java/com/apollographql/apollo/internal/fetcher/NetworkFirstFetcher.java

```diff
@@ -38,12 +38,28 @@ public void interceptAsync(@Nonnull final InterceptorRequest request, @Nonnull f
           callBack.onResponse(response);
         }
 
-        @Override public void onFailure(@Nonnull ApolloException e) {
-          logger.d(e, "Failed to fetch network response for operation %s, trying to return cached one",
+        @Override public void onFailure(@Nonnull final ApolloException networkException) {
+          logger.d(networkException, "Failed to fetch network response for operation %s, trying to return cached one",
               request.operation);
           if (!disposed) {
             InterceptorRequest cacheRequest = request.toBuilder().fetchFromCache(true).build();
-            chain.proceedAsync(cacheRequest, dispatcher, callBack);
+            chain.proceedAsync(cacheRequest, dispatcher,  new CallBack() {
+              @Override public void onResponse(@Nonnull InterceptorResponse response) {
+                callBack.onResponse(response);
+              }
+
+              @Override public void onFetch(FetchSourceType sourceType) {
+                callBack.onFetch(sourceType);
+              }
+
+              @Override public void onFailure(@Nonnull ApolloException cacheException) {
+                callBack.onFailure(networkException);
+              }
+
+              @Override public void onCompleted() {
+                callBack.onCompleted();
+              }
+            });
```

Comment:
I guess we need to check if it wasn't disposed before notifying original callback, like:
```
if (!disposed) {
callBack.onFailure(networkException);
}
```

What do you think?